_drafts/release-workflows.md

```diff
@@ -0,0 +1,383 @@
+
+| Action                                                | Result                                                                                                                                                                                                                                                                                                    |
+|-------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| **push** to `master`                                  | Release on `latest` channel if the commit trigger a release with a version that doesn't satisfies the `type` defines for `next` branch, report an error otherwise                                                                                                                                         |
+| **merge** feature branch to `master`                  | Release on `latest` channel if the commit trigger a release with a version that doesn't satisfies the `type` defines for `next` branch, report an error otherwise                                                                                                                                         |
+| **push** to `next`                                    | Release on `next` channel                                                                                                                                                                                                                                                                                 |
+| **merge** any branch to `next`                        | Release on `next` channel                                                                                                                                                                                                                                                                                 |
+| **merge** some commits from `next` branch to `master` | Get the last release for commits on branch `master` (including the one that come from `next`), analyze commits, increase the version; If the commit associated with the last release found is also present on `next`, make the version available on the `latest` channel, otherwise do a regular release. |
+
+##### Future branches configuration
+
+```json
+"release": {
+  "branches": ["master", {
+		"branch": "next",
+		"type": "major"
+		}]
+}
+```
+
+The `type` is required for future branches and indicate that its reserved for a certain type of release (`minor` or `major`).
+That defines how much ahead the future branch must be versus the pervious branch in the list. If configured with `major`, the `next` branch must always one `major` release ahead of `master`.
```

Comment:
> I wonder if we need the type configuration? Maybe we can launch without it and see if situations come up where such a setting would become necessary?

I didn't had the `type` property at the beginning but that would create inconsistent situation that would be impossible to detect by semantic-release and from which it would really difficult to recover.
The concept of `type` is necessary to avoid situation like this:
- `master` with `1.0.0` being the last release done from there
- `next` created from `master`
- `experimental` created from `master`

If you push a breaking change on `next` that would release `2.0.0` on `next` dist-tag. If you push another breaking change to `experimental` that would release `2.0.0` on `experimental` dist-tag which is impossible. npm wouldn't allow it, and it's inconsistent anyway as you can't have two different releases with the same version.

We need the concept of `type` to ensure that each type of releases (patch, minor/feature, major/breaking) is allowed only from one branch.

> I’m not sure if defaulting type to major is the best idea. I probably will use next most often for breaking version bumps, but not always. In some cases there is a new feature that I want to release as next first. I think semantic-release could intelligently figure out what versions are allowed in either branch based on current releases?

`type` represent the release gap allowed with the default branch. It must be a valid Semver release type: `patch`, `minor` or `major`. Defaulting to `major` means that by default we allow any type of releases on a Future branch (which what you describe in your example as your `next` branch).

Basically it means there is no release gap restriction between `next` and `master`, you can release whatever you want on `next`. Only the normal consistency restriction apply: if `master` gets ahead of `next` it must contains all the commits from `next`.
It has for consequence that:
- if you use `next` for a breaking release, you can do only `fix` and `feat` on `master`
- if you use `next` for a `feat`, you can do  only `fix` on `master`

In other worlds:
- What you describe as " intelligently figure out what versions are allowed" is the default behavior
- Setting a `type` manually allow users to be more restrictive and impose certain releases in advance before there is enough commits available for semantic-release to figure it out (something like that: https://github.com/sass/sass/blob/stable/CONTRIBUTING.md#branches)

Reply:
> If you push another breaking change to experimental that would release 2.0.0 on experimental dist-tag which is impossible

I can’t think of a use case where I would want releases happening from an `experimental` branch at all, my assumption is that there is one branch for `latest` releases and one branch for future releases which I’d call `next`. 

I know there are cases to go beyond it but I feel we should do this first and then look into support for things like `5.0.0-beta.1` once we have the basics right.

Reply:
I know that getting this `latest` & `next` branch setup with automated releases for the different channels is hard, but I think if we manage to figure this out, it will have big potential for a great developer experience

Reply:
This entire spec is based on using the power of Semver. What you propose it to make more complex code to limit features.
If we go with the `next` branch only we would have to do custom code that. When we want to do all the features described here, we would have to throw away this custom code to replace it by more abstract code that use semver.
I don't understand the concept of doing the work twice (first time specific with `next`, second time abstract using semver).

> I know that getting this latest & next branch setup with automated releases for the different channels is hard, but I think if we manage to figure this out, it will have big potential for a great developer experience

That's exactly what this spec is doing, and that's why it's complex. Taht what I was trying to explain in my previous comment. I really don't see what is missing there....